atc/exec/task_step.go

```diff
@@ -188,135 +184,69 @@ func (step *TaskStep) Run(ctx context.Context, state RunState) error {
-		ctx,
-		logger,
-		step.delegate,
-		owner,
-		step.containerMetadata,
-		containerSpec,
-		resourceTypes,
-	)
-	if err != nil {
-		return err
+	processSpec := worker.TaskProcessSpec{
+		Path:         config.Run.Path,
+		Args:         config.Run.Args,
+		Dir:          config.Run.Dir,
+		StdoutWriter: step.delegate.Stdout(),
+		StderrWriter: step.delegate.Stderr(),
 	}
 
-	exitStatusProp, err := container.Property(taskExitStatusPropertyName)
-	if err == nil {
-		logger.Info("already-exited", lager.Data{"status": exitStatusProp})
-
-		status, err := strconv.Atoi(exitStatusProp)
-		if err != nil {
-			return err
-		}
-
-		step.succeeded = (status == 0)
-
-		err = step.registerOutputs(logger, repository, config, container, step.containerMetadata)
-		if err != nil {
-			return err
-		}
-
-		return nil
-	}
-
-	processIO := garden.ProcessIO{
-		Stdout: step.delegate.Stdout(),
-		Stderr: step.delegate.Stderr(),
+	imageSpec := worker.ImageFetcherSpec{
+		ResourceTypes: resourceTypes,
+		Delegate:      step.delegate,
 	}
+	owner := db.NewBuildStepContainerOwner(step.metadata.BuildID, step.planID, step.metadata.TeamID)
 
-	process, err := container.Attach(taskProcessID, processIO)
-	if err == nil {
-		logger.Info("already-running")
-	} else {
-		logger.Info("spawning")
-
-		step.delegate.Starting(logger, config)
-
-		process, err = container.Run(
-			garden.ProcessSpec{
-				ID: taskProcessID,
-
-				Path: config.Run.Path,
-				Args: config.Run.Args,
-
-				Dir: path.Join(step.containerMetadata.WorkingDirectory, config.Run.Dir),
-
-				// Guardian sets the default TTY window size to width: 80, height: 24,
-				// which creates ANSI control sequences that do not work with other window sizes
-				TTY: &garden.TTYSpec{
-					WindowSize: &garden.WindowSize{Columns: 500, Rows: 500},
-				},
-			},
-			processIO,
+	results := make(chan worker.TaskResult)
+	events := make(chan string, 1)
+
+	go func(results chan worker.TaskResult) {
+		results <- step.workerClient.RunTaskStep(
+			ctx,
+			logger,
+			owner,
+			containerSpec,
+			workerSpec,
+			step.strategy,
+			step.containerMetadata,
+			imageSpec,
+			processSpec,
+			events,
 		)
-	}
-	if err != nil {
-		return err
-	}
-
-	logger.Info("attached")
-
-	exited := make(chan struct{})
-	var processStatus int
-	var processErr error
-
-	go func() {
-		processStatus, processErr = process.Wait()
-		close(exited)
-	}()
-
-	select {
-	case <-ctx.Done():
-		err = step.registerOutputs(logger, repository, config, container, step.containerMetadata)
-		if err != nil {
-			return err
-		}
-
-		err = container.Stop(false)
-		if err != nil {
-			logger.Error("stopping-container", err)
-		}
+	}(results)
+
+	for {
+		select {
+		case <-ctx.Done():
+			result := <-results
+			err = step.registerOutputs(logger, repository, config, result.VolumeMounts, step.containerMetadata)
+			if err != nil {
+				return err
+			}
 
-		<-exited
+			return ctx.Err()
 
-		return ctx.Err()
+		case result := <-results:
```

Comment:
Can we unify this code path with the ctx.Done code path? 

I'm a little confused why we need both. [In the old task step](https://github.com/concourse/concourse/blob/master/atc/exec/task_step.go#L284-L318) it seemed really clear that we were waiting on either the [process to exit](https://github.com/concourse/concourse/blob/master/atc/exec/task_step.go#L299), or the [context to be cancelled](https://github.com/concourse/concourse/blob/master/atc/exec/task_step.go#L284) (which would prompt us to [stop the container](https://github.com/concourse/concourse/blob/master/atc/exec/task_step.go#L290) and then still wait for the [process to exit](https://github.com/concourse/concourse/blob/master/atc/exec/task_step.go#L295)).

It looks like this logic still exists in `RunTaskStep`, so just wondering why we need both here?

Reply:
As per our discussion, we changed event handling to be handled in a separate goroutine. This allowed us to collapse the `select` block in question into just one code block (no channels required). Instead, event handling is the only thing requiring a channel now.